docs/proposals/20200615-external-remediation.md

```diff
@@ -0,0 +1,299 @@
+
+#### Proposed CRD changes and additions
+
+    type MachineHealthCheckSpec struct {
+        ...
+    
+        // +optional
+        RemediationTemplate ObjectReference `json:"remediationTemplate,omitempty"`
+    }
+
+#### Example CRs
+
+MachineHealthCheck:
+    
+    kind: MachineHealthCheck
+    apiVersion: cluster.x-k8s.io/v1alphaX
+    metadata:
+      name: REMEDIATION_GROUP
+      namespace: NAMESPACE_OF_UNHEALTHY_MACHINE
+    spec:
+      selector:
+        matchLabels: 
+          ...
+      remediationTemplate:
+        kind: Metal3RemediationTemplate
+        apiVersion: remediation.metal3.io/v1alphaX
+        name: M3_REMEDIATION_GROUP
+
+Metal3RemediationTemplate:
+
+    kind: Metal3RemediationTemplate
+    apiVersion: remediation.metal3.io/v1alphaX
+    metadata:
+      name: M3_REMEDIATION_GROUP
+      namespace: NAMESPACE_OF_UNHEALTHY_MACHINE
+    spec:
+      template:
+        spec:
+          strategy: 		    escalate
+          deleteAfterRetries:     10
+          powerOnTimeoutSeconds:  600
+          powerOffTimeoutSeconds: 120
+
+Metal3Remediation:
+
+    apiVersion: remediation.metal3.io/v1alphaX
+    kind: Metal3Remediation
+    metadata:
+      name: NAME_OF_UNHEALTHY_MACHINE
+      namespace: NAMESPACE_OF_UNHEALTHY_MACHINE
+       finalizer: 
+       - remediation.metal3.io
+       ownerReferences: 
+    apiVersion:cluster.x-k8s.io/v1alphaX
+    kind: MachineHealthCheck
+    name: REMEDIATION_GROUP
+    uid: ...
+      spec: 
+        strategy: 		    escalate
+        deleteAfterRetries:     10
+        powerOnTimeoutSeconds:  600
+        powerOffTimeoutSeconds: 120
+      status:
+        phase: power off
+        retryCount: 1
+
+#### Usage
+
+    
+    func (t *healthCheckTarget) createProviderRemediation(ctx context.Context) (*corev1.ObjectReference, error) {
+        if t.MHC.Spec.RemediationTemplate == nil {
+            return nil, nil
+    }
+    
+    from, err := Get(ctx, c, t.MHC.Spec.RemediationTemplate, t.Machine.Namespace)
+    if err != nil {
+        return nil, err
+    }
+    generateTemplateInput := &GenerateTemplateInput{
+        Template:    from,
+        Namespace:   t.Machine.Namespace,
+        ClusterName: t.Machine.ClusterName,
+        OwnerRef:    metav1.OwnerReference{*metav1.NewControllerRef(
+    t.Machine, t.Machine.Kind)},
+        Labels:      [],
+    }
+    to, err := GenerateTemplate(generateTemplateInput)
+    if err != nil {
+        return nil, err
+    }
+    
+    // Force the same name as the machine
+    to.SetName(t.Machine.Name)
+    
+    // Create the external clone.
+    if err := r.Client.Create(context.Background(), to); err != nil {
+        return nil, err
+    }
+    return GetObjectReference(to), nil
+    }
+    
+    func (t *healthCheckTarget) remediate(...) error {
+        ...
+    remediation, err := t.createProviderRemediation(ctx)
```

Comment:
Before I review the code suggestions any further, how would you feel about removing them from the proposal, and we'd defer reviewing the code until there is a PR?

Reply:
That make sense.

Reply:
Sure, I'll remove the code. thanks